drivers/regulator/regulator_da1469x.c

```diff
@@ -0,0 +1,451 @@
+}
+
+static int regulator_da1469x_list_voltage(const struct device *dev,
+					  unsigned int idx,
+					  int32_t *volt_uv)
+{
+	const struct regulator_da1469x_config *config = dev->config;
+
+	if (idx >= config->voltages_count) {
+		return -EINVAL;
+	}
+
+	*volt_uv = config->voltages[idx];
+
+	return 0;
+}
+
+static int regulator_da1469x_set_voltage(const struct device *dev, int32_t min_uv,
+					 int32_t max_uv)
+{
+	int ret;
+	const struct regulator_da1469x_config *config = dev->config;
+	int i;
+
+	/* Check if voltage can be provided */
+	for (i = 0; i < config->voltages_count; ++i) {
+		if (min_uv <= config->voltages[i] && config->voltages[i] <= max_uv) {
+			break;
+		}
+	}
+	if (i < config->voltages_count) {
+		ret = 0;
+		switch (config->rail) {
+		case VDD_CLAMP:
+			POWER_REG_SET(VDD_CLAMP_LEVEL, i);
+			break;
+		case VDD_SLEEP:
+			POWER_REG_SET(VDD_SLEEP_LEVEL, i);
+			break;
+		case VDD:
+			POWER_REG_SET(VDD_LEVEL, i);
+			break;
+		case V14:
+			POWER_REG_SET(V14_LEVEL, i);
+			break;
+		case V18:
+			POWER_REG_SET(V18_LEVEL, i);
+			break;
+		case V18P:
+			break;
+		case V30:
+			POWER_REG_SET(V30_LEVEL, i << 1);
+			break;
+		}
+	} else {
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+static int regulator_da1469x_get_voltage(const struct device *dev,
+					 int32_t *volt_uv)
+{
+	const struct regulator_da1469x_config *config = dev->config;
+	int voltage_ix;
+
+	switch (config->rail) {
+	case VDD_CLAMP:
+		voltage_ix = POWER_REG_GET(VDD_CLAMP_LEVEL);
+		break;
+	case VDD:
+		voltage_ix = POWER_REG_GET(VDD_LEVEL);
+		break;
+	case VDD_SLEEP:
+		voltage_ix = POWER_REG_GET(VDD_SLEEP_LEVEL);
+		break;
+	case V18:
+		voltage_ix = POWER_REG_GET(V18_LEVEL);
+		break;
+	case V14:
+		voltage_ix = POWER_REG_GET(V14_LEVEL);
+		break;
+	case V30:
+		voltage_ix = POWER_REG_GET(V30_LEVEL) >> 1;
+		break;
+	default:
+		voltage_ix = 0;
+		break;
+	}
+	*volt_uv = config->voltages[voltage_ix];
+
+	return 0;
+}
+
+static int regulator_da1469x_set_current_limit(const struct device *dev,
+					       int32_t min_ua, int32_t max_ua)
+{
+	const struct regulator_da1469x_config *config = dev->config;
+	int ret = 0;
+
+	switch (config->rail) {
+	case VDD:
+		DCDC_REG_SET(VDD, CUR_LIM_MAX_HV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(VDD, CUR_LIM_MAX_LV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(VDD, CUR_LIM_MIN, ((max_ua / 30000) - 1));
+		break;
+	case V14:
+		DCDC_REG_SET(V14, CUR_LIM_MAX_HV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V14, CUR_LIM_MAX_LV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V14, CUR_LIM_MIN, ((max_ua / 30000) - 1));
+		break;
+	case V18:
+		DCDC_REG_SET(V18, CUR_LIM_MAX_HV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V18, CUR_LIM_MAX_LV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V18, CUR_LIM_MIN, ((max_ua / 30000) - 1));
+		break;
+	case V18P:
+		DCDC_REG_SET(V18P, CUR_LIM_MAX_HV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V18P, CUR_LIM_MAX_LV, ((max_ua / 30000) - 1));
+		DCDC_REG_SET(V18P, CUR_LIM_MIN, ((max_ua / 30000) - 1));
+		break;
```

Comment:
magic numbers here, remember current is given in a range as well

Reply:
changed to linear ranges